File: packages/core/src/mastering/dyneq.py
"""Linked-detection dynamic EQ.

The profile EQ is a static curve and the bus compressor is full-range, so
neither can do the surgical moves mastering actually reaches for: tame a harsh
3-4 kHz region only when it flares, tuck a resonant low-mid on loud passages.
Each band here is a parametric bell that rides between 0 dB and whatever its
own detector asks for, so the signal is never split into bands and recombined
— below its threshold a band is the identity, sample for sample.

Detection is per band and **linked across channels**: one band-pass per bed
channel feeds a single RMS, exactly as ``bus_compress``'s sidechain does, and
the resulting gain is realized as one bell design applied identically to every
non-LFE channel.  That keeps the stage a shared time-varying filter, which
commutes with the LF sum the same way the linked compressor's gain does
(``docs/contracts/preview_export_parity.md`` §1), and makes channel divergence
— the failure that closed ``docs/plans/mixing/phase13_report.md`` — impossible
by construction.

The gain computer is the compressor's, so ``threshold_db`` and ``ratio`` mean
what they mean there; the soft knee is structural and lives in
``mastering::dyneq``'s ``KNEE_DB``.
"""
from __future__ import annotations

import logging

import numpy as np
import upmixer_dsp

_log = logging.getLogger("upmixer")

MANIFEST_FIELDS = {
        "profile": ("config", "mastering_dyneq_profile"),
        "bands":   ("config", "mastering_dyneq_bands"),
}

BAND_FIELDS: tuple[str, ...] = (
    "freq_hz",
    "q",
    "threshold_db",
    "ratio",
    "attack_ms",
    "release_ms",
)

#: Bands a chain may carry, mirroring ``mastering::dyneq``'s ``MAX_BANDS``.
MAX_BANDS = 4

DYNEQ_PROFILES: dict[str, list[dict]] = {
    "tame-harshness": [
        dict(freq_hz=3500.0, q=1.8, threshold_db=-32.0, ratio=3.0,
             attack_ms=15.0, release_ms=180.0),
    ],
    "tame-sibilance": [
        dict(freq_hz=7500.0, q=3.0, threshold_db=-31.0, ratio=4.0,
             attack_ms=2.0, release_ms=80.0),
    ],
    "clear-low-mid": [
        dict(freq_hz=250.0, q=1.2, threshold_db=-25.0, ratio=2.5,
             attack_ms=30.0, release_ms=250.0),
    ],
    "tighten-low-end": [
        dict(freq_hz=75.0, q=1.0, threshold_db=-23.0, ratio=3.0,
             attack_ms=20.0, release_ms=200.0),
    ],
    "immersive-polish": [
        dict(freq_hz=250.0, q=1.2, threshold_db=-25.0, ratio=2.0,
             attack_ms=30.0, release_ms=250.0),
        dict(freq_hz=3500.0, q=1.8, threshold_db=-31.0, ratio=2.5,
             attack_ms=15.0, release_ms=180.0),
        dict(freq_hz=11000.0, q=2.0, threshold_db=-29.0, ratio=2.5,
             attack_ms=5.0, release_ms=120.0),
    ],
}
"""Named band sets, chosen for what *this* pipeline does to a bed rather than
for generic mastering moves.

``clear-low-mid`` and ``tighten-low-end`` target coherent summing: bass
management unifies below its crossover, but from there to ~400 Hz every bed
channel still carries correlated content that sums at up to 10·log₁₀(N) dB, and
nothing else in the chain addresses it.  ``tame-harshness`` and
``immersive-polish``'s middle band target the presence region, which the height
sends' high shelf and the surround/height decorrelation both push on.
``tame-sibilance`` is the one bus-level move the knowledge base argues against
(``techniques/mastering_restoration.md``: de-ess the vocal stem instead) — it
is here because the realtime pipeline has no stems to fix at stem level, and in
stem mode it is a safety net rather than the first tool.

Thresholds are absolute dBFS on the *pre*-normalization bed, the same
convention and the same chain position as :data:`COMP_PROFILES`.  They are
calibrated for a bed whose full-band linked RMS sits near −20 dBFS, which is
what those profiles also assume.

The two families are set differently because they catch different things.  The
high bands sit well above their band's median, near where a dense programme's
transient flares reach, so they act on those and are inert between them.  The
low bands sit near their median: correlated low-frequency content varies by
under a dB within a passage, so there are no flares to catch there — what a low
band is for is the 6-10 dB a mix moves between a quiet passage and a loud one,
which means engaging through the loud one and relaxing through the quiet one.
"""

DYNEQ_PROFILE_NAMES: tuple[str, ...] = tuple(sorted(DYNEQ_PROFILES))


def resolve_dyneq_bands(
    profile: str | None,
    bands: list[dict] | None,
) -> list[dict]:
    """Bands the chain should run: explicit *bands* win over *profile*.

    Manifest values beat profile defaults, the same precedence every other
    stage applies.  An unknown profile name resolves to no bands.
    """
    if bands:
        return bands
    if profile is None:
        return []
    return [dict(band) for band in DYNEQ_PROFILES.get(profile, ())]


def apply_dynamic_eq(
    channels: dict[str, np.ndarray],
    sample_rate: int,
    bands: list[dict],
    lfe_key: str = "LFE",
    detector_channels: dict[str, np.ndarray] | None = None,
    detector_lfe_key: str = "LFE",
) -> dict[str, np.ndarray]:
    """Run every band over the bed, leaving *lfe_key* untouched.

    Args:
        channels:    Dict channel_name -> 1-D float array.
        sample_rate: Audio sample rate in Hz.
        bands:       One dict per band, keyed by :data:`BAND_FIELDS`.
        lfe_key:     Channel name kept out of both the detector and the bells.
        detector_channels: Optional rendered speaker programme that drives the
            shared bell curves applied to ``channels``.
        detector_lfe_key: LFE name in ``detector_channels``.

    Returns:
        New channel dict with the same shapes and dtypes.
    """
    if not channels or not bands:
        return channels
    names = list(channels)
    targets = [np.ascontiguousarray(channels[name], dtype=np.float64) for name in names]
    params = [(*tuple(float(band[field]) for field in BAND_FIELDS[:4]),
               float(band.get("max_cut_db", float("inf"))),
               *tuple(float(band[field]) for field in BAND_FIELDS[4:])) for band in bands]
    if detector_channels is None:
        processed, cuts = upmixer_dsp.dynamic_eq(
            targets,
            names.index(lfe_key) if lfe_key in channels else None,
            sample_rate,
            params,
        )
    else:
        detector_names = list(detector_channels)
        processed, cuts = upmixer_dsp.dynamic_eq_linked(
            targets,
            names.index(lfe_key) if lfe_key in channels else None,
            [
                np.ascontiguousarray(detector_channels[name], dtype=np.float64)
                for name in detector_names
            ],
            detector_names.index(detector_lfe_key)
            if detector_lfe_key in detector_channels else None,
            sample_rate,
            params,
        )
    for band, cut in zip(bands, cuts):
        _log.info(
            "  Dynamic EQ: %.0f Hz Q %.1f  peak cut %.2f dB",
            band["freq_hz"], band["q"], cut,
        )
    return {
        name: arr.astype(channels[name].dtype) for name, arr in zip(names, processed)
    }

File: packages/dsp/crates/dsp-core/src/mastering/dyneq.rs
//! Linked-detection dynamic EQ: a few bell bands that act only when their own
//! band flares.
//!
//! Each band carries one detector per bed channel — the band's own RBJ
//! band-pass, built from the same `(f0, Q)` as its bell — and sums their
//! powers into a single linked RMS, the topology
//! [`super::compressor::bus_compress`] uses. That produces *one* gain
//! trajectory per band, realized as one bell design applied identically to
//! every non-LFE channel, so the stage stays a shared time-varying filter
//! across the bed and commutes with the LF sum (see
//! `docs/contracts/preview_export_parity.md` §1).
//!
//! Detection is per band but never per channel: the failure that closed
//! mixing phase 13 was independent detectors diverging across a broadband
//! decay, and nothing here can diverge across channels by construction.
//!
//! The bell's coefficients are recomputed whenever its gain moves, keeping
//! the section's delay registers, and at unity gain the design is exactly
//! `b == a`, so a band that never crosses its threshold returns the input
//! sample for sample.

use crate::kernels::biquad::{bandpass_sos, peaking_sos, Sos};

use super::compressor::{alpha, knee_gain_db};
use super::non_lfe;

/// Soft knee every band's gain computer runs, in dB. Structural rather than
/// tunable: a hard knee chatters as the detector crosses the threshold, and
/// the value only has to be wide enough to stop that.
pub const KNEE_DB: f64 = 6.0;

/// Bands a chain may carry. Surgical correction is a handful of moves, and
/// each band costs two biquads per channel per sample on the audio thread.
pub const MAX_BANDS: usize = 4;

#[derive(Clone, Copy, Debug, PartialEq, serde::Deserialize)]
pub struct BandParams {
    pub freq_hz: f64,
    pub q: f64,
    /// Level the band's own linked RMS has to reach before the bell engages.
    pub threshold_db: f64,
    /// Slope above the threshold, exactly as the bus compressor's ratio —
    /// 1.0 leaves the band inert whatever the detector reads.
    pub ratio: f64,
    pub attack_ms: f64,
    pub release_ms: f64,
}

struct Band {
    params: BandParams,
    /// One detector and one bell per bed channel, in `channels` order.
    detect: Vec<Sos>,
    bell: Vec<Sos>,
    attack: f64,
    release: f64,
    fast: f64,
    slow: f64,
    /// Design constants of the bell that do not depend on its gain.
    alpha: f64,
    cos_w0: f64,
    /// Gain the bell sections currently realize, so the design is rebuilt
    /// only when the detector actually moves it.
    gain: f64,
    peak_cut_db: f64,
}

impl Band {
    fn new(params: BandParams, sample_rate: u32, n_bed: usize) -> Self {
        let nyq = sample_rate as f64 / 2.0;
        let wn = (params.freq_hz / nyq).clamp(1e-4, 0.999);
        let q = params.q.max(0.1);
        let w0 = std::f64::consts::PI * wn;
        Self {
            detect: vec![Sos::new(bandpass_sos(wn, q)); n_bed],
            bell: vec![Sos::new(peaking_sos(wn, q, 1.0)); n_bed],
            attack: alpha(params.attack_ms, sample_rate),
            release: alpha(params.release_ms, sample_rate),
            fast: 0.0,
            slow: 0.0,
            alpha: w0.sin() / (2.0 * q),
            cos_w0: w0.cos(),
            gain: 1.0,
            peak_cut_db: 0.0,
            params,
        }
    }

    fn reset(&mut self) {
        for section in self.detect.iter_mut().chain(self.bell.iter_mut()) {
            section.reset();
        }
        self.fast = 0.0;
        self.slow = 0.0;
        self.peak_cut_db = 0.0;
    }

    /// Detector RMS across the bed for one sample, advancing every band-pass.
    #[inline]
    fn linked_rms(&mut self, bed: &super::Bed, channels: &[usize], frame: usize) -> f64 {
        let mut acc = 0.0;
        for (slot, &channel) in channels.iter().enumerate() {
            let v = self.detect[slot].tick(bed[channel][frame]);
            acc += v * v;
        }
        (acc / channels.len() as f64 + 1e-20).sqrt()
    }

    /// Gain the bell should realize for one sample, from the detector level.
    #[inline]
    fn gain_for(&mut self, rms: f64) -> f64 {
        self.fast += self.attack * (rms - self.fast);
        self.slow += self.release * (rms - self.slow);
        let env_db = 20.0 * self.fast.max(self.slow).max(1e-20).log10();
        let cut_db = knee_gain_db(env_db, self.params.threshold_db, self.params.ratio, KNEE_DB);
        self.peak_cut_db = self.peak_cut_db.max(-cut_db);
        10.0_f64.powf(cut_db / 20.0)
    }

    /// Filter one sample of every bed channel through the bell, redesigning
    /// it first when the gain moved.
    #[inline]
    fn apply(&mut self, bed: &mut super::Bed, channels: &[usize], frame: usize, gain: f64) {
        if gain != self.gain {
            let a = gain.sqrt();
            let coeffs = [
                1.0 + self.alpha * a,
                -2.0 * self.cos_w0,
                1.0 - self.alpha * a,
                1.0 + self.alpha / a,
                -2.0 * self.cos_w0,
                1.0 - self.alpha / a,
            ];
            for section in &mut self.bell {
                section.retune(coeffs);
            }
            self.gain = gain;
        }
        for (slot, &channel) in channels.iter().enumerate() {
            bed[channel][frame] = self.bell[slot].tick(bed[channel][frame]);
        }
    }
}

/// The stage: every band, sharing one bed-channel list.
pub struct DynamicEq {
    bands: Vec<Band>,
    channels: Vec<usize>,
}

impl DynamicEq {
    /// `None` when nothing would act — no bands, no bed, or every band at a
    /// ratio of 1.0 — so the stage costs nothing when it is off.
    pub fn new(
        sample_rate: u32,
        n_channels: usize,
        lfe: Option<usize>,
        bands: &[BandParams],
    ) -> Option<Self> {
        let channels = non_lfe(n_channels, lfe);
        let bands: Vec<Band> = bands
            .iter()
            .filter(|b| b.ratio > 1.0)
            .map(|b| Band::new(*b, sample_rate, channels.len()))
            .collect();
        (!bands.is_empty() && !channels.is_empty()).then_some(Self { bands, channels })
    }

    pub fn reset(&mut self) {
        for band in &mut self.bands {
            band.reset();
        }
    }

    /// Whether this stage was built for exactly `bands`, so a live parameter
    /// edit that changed nothing does not restart the detectors.
    pub fn matches(&self, bands: &[BandParams]) -> bool {
        let active: Vec<&BandParams> = bands.iter().filter(|b| b.ratio > 1.0).collect();
        self.bands.len() == active.len()
            && self.bands.iter().zip(active).all(|(band, p)| band.params == *p)
    }

    /// Run every band over the whole bed, in place.
    pub fn process(&mut self, bed: &mut super::Bed) {
        let frames = self.channels.iter().map(|&i| bed[i].len()).min().unwrap_or(0);
        for frame in 0..frames {
            for band in &mut self.bands {
                let rms = band.linked_rms(bed, &self.channels, frame);
                let gain = band.gain_for(rms);
                band.apply(bed, &self.channels, frame, gain);
            }
        }
    }

    /// Deepest cut each band reached, in dB, in the order they were given.
    pub fn peak_cut_db(&self) -> Vec<f64> {
        self.bands.iter().map(|b| b.peak_cut_db).collect()
    }
}

/// Run the dynamic EQ over a whole bed, returning each band's deepest cut.
pub fn dynamic_eq(
    bed: &mut super::Bed,
    lfe: Option<usize>,
    sample_rate: u32,
    bands: &[BandParams],
) -> Vec<f64> {
    match DynamicEq::new(sample_rate, bed.len(), lfe, bands) {
        Some(mut stage) => {
            stage.process(bed);
            stage.peak_cut_db()
        }
        None => Vec::new(),
    }
}
